src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingRouteAdapter.cs

```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Globalization;
+using System.Linq;
+using Grpc.Shared;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Grpc.JsonTranscoding.Internal;
+
+/// <summary>
+/// Routes on HTTP rule are similar to ASP.NET Core routes but add and remove some features.
+/// - Constraints aren't supported.
+/// - Optional parameters aren't supported.
+/// - Parameters spanning multiple segments are supported.
+///
+/// This type transforms an HTTP route into an ASP.NET Core route by rewritting it to a compatible
+/// format and providing actions to reconstruct parameters that span multiple segments.
+/// </summary>
+internal class JsonTranscodingRouteAdapter
+{
+    public HttpRoutePattern HttpRoutePattern { get; }
+    public string ResolvedRouteTemplate { get; }
+    public List<Action<HttpContext>> RewriteVariableActions { get; }
+
+    private JsonTranscodingRouteAdapter(HttpRoutePattern httpRoutePattern, string resolvedRoutePattern, List<Action<HttpContext>> rewriteVariableActions)
+    {
+        HttpRoutePattern = httpRoutePattern;
+        ResolvedRouteTemplate = resolvedRoutePattern;
+        RewriteVariableActions = rewriteVariableActions;
+    }
+
+    public static JsonTranscodingRouteAdapter Parse(HttpRoutePattern pattern)
```

Comment:
Performance?

Reply:
This runs for each route on startup and is then cached. Requests reuse the adapter.

It's not performance-critical.

Reply:
Oh, and the adapted routes passed to ASP.NET Core still use DFA route matching. e.g. `/v1/{book=shelves/*/books/*}` turns into `/v1/shelves/{param1}/books/{param2}`. As fast as you can get.

It doesn't do something inefficient like map a catch-all route and then reimplement matching in the gRPC JSON transcoding library.